### PR TITLE
FIX Archive record from live table on delete

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-linkfield/issues/213

### CODE REVIEW

File: tests/php/Controllers/LinkFieldControllerTest.php [hunk 1]
```diff
@@ -500,7 +513,6 @@ public function testLinkDelete(
             $owner = $this->getFixtureLinkOwner();
             $this->assertSame(0, $owner->LinkID);
         }
-        $this->assertTrue(true);
```
Comment (emteknetnz) [inline tests/php/Controllers/LinkFieldControllerTest.php:503]: This is unrelated refactoring to remove a pointless assertion

File: src/Controllers/LinkFieldController.php [hunk 2]
```diff
@@ -136,15 +137,28 @@ public function linkDelete(): HTTPResponse
...
+            // Update draft owner table
             $owner->$ownerRelation = null;
             $owner->write();
+            // Update live owner table if owner is versioned
+            if ($owner->hasExtension(Versioned::class)) {
+                Versioned::withVersionedMode(function () use ($ownerRelation) {
+                    Versioned::set_stage(Versioned::LIVE);
+                    $liveOwner = $this->getOwnerFromRequest();
+                    $liveOwner->$ownerRelation = null;
+                    $liveOwner->write();
```
Comment (emteknetnz) [inline src/Controllers/LinkFieldController.php:159]: I'm not sure if we should use write() here, or should just do something closer to raw sql that bypasses the ORM

File: src/Controllers/LinkFieldController.php [hunk 3]
```diff
@@ -136,15 +137,28 @@ public function linkDelete(): HTTPResponse
...
+                });
+            }
```
Comment (ChloeSartorelli) [inline src/Controllers/LinkFieldController.php]: This incorrectly published changes which are in draft.
To reproduce:
1. Create a link
2. Publish the page
3. Make some content change on the page and save (but don't publish)
4. Archive the link

Reloading the page will show a published state for the page (i.e. publish button is not green).
Visiting the history will show the published state has the draft changes in it.
HOWEVER! Viewing the page in the front-end shows the old published content (not the draft content that got sort-of-maybe-published?)

Not sure if that's a bug with versioned or if this is just being done wrong or what, but something's not right.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Controllers/LinkFieldController.php (hunk 3 above)
Comment (emteknetnz) [reply src/Controllers/LinkFieldController.php]: Updated to use a SQLUpdate instead

File: src/Controllers/LinkFieldController.php [hunk 4]
```diff
@@ -136,15 +138,29 @@ public function linkDelete(): HTTPResponse
...
             $owner->write();
+            // Update live owner table if owner is versioned
+            if ($owner->hasExtension(Versioned::class)) {
+                $draftTable = $owner->getSchema()->tableForField($owner::class, $ownerRelation . 'ID');
+                $liveTable = $draftTable . '_Live';
+                SQLUpdate::create()
+                    ->setTable($liveTable)
+                    ->assign('"' . $ownerRelation . 'ID"', 0)
+                    ->addWhere(['ID' => $owner->ID])
+                    ->execute();
```
Comment (ChloeSartorelli) [inline src/Controllers/LinkFieldController.php]: This feels a bit hacky.
It's also not consistent with what happens when you archive related records in other situations. I don't think we should be doing this part.

The ORM already knows how to handle the scenario where a related record doesn't exist anymore. We don't need to try to handle that here.

If the old ID is causing errors in the UI, we need to instead swallow those errors (an outdated ID is fine, we just treat is as though the ID wasn't there in literally every other scenario in the CMS)

Comment (ChloeSartorelli) [reply src/Controllers/LinkFieldController.php]: Should have realised this last time, sorry about that.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

**Review by GuySartorelli — APPROVED** [review]
Looks great, works well locally.

There's some bad toasts when looking in the history tab, but that can be handled in https://github.com/silverstripe/silverstripe-linkfield/issues/10

**Pull request merged**